### PR TITLE
[ESIMD] Infer address space of pointer that are passed through invoke_simd to ESIMD API to generate better code on BE

### PR DESCRIPTION
**Pull request by fineg74**

### CODE REVIEW

File: llvm/lib/SYCLLowerIR/LowerInvokeSimd.cpp [hunk 1]
```diff
@@ -259,6 +262,49 @@ void markFunctionAsESIMD(Function *F) {
...
+
+      for (uint32_t i = 0; i < CI->getNumOperands(); ++i) {
+        if (CI->getOperand(i) == Arg) {
+          AdjustAddressSpace(Callee, i, ArgAddrSpace);
+        }
+      }
+    } else {
+      for (unsigned int i = 0; i < ArgUse->getNumOperands(); ++i) {
+        if (ArgUse->getOperand(i) == Arg) {
+          const Type *Ty = ArgUse->getOperand(i)->getType();
```
Comment (sarnex) [inline llvm/lib/SYCLLowerIR/LowerInvokeSimd.cpp:293]: It looks like we only use this variable to get the context, probably we could get rid of it and get the context from somewhere else

File: sycl/test/invoke_simd/invoke_simd_address_space_inferral.cpp [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+  } catch (sycl::exception const &e) {
+    std::cout << "SYCL exception caught: " << e.what() << '\n';
+    sycl::free(A, q);
+    return false;
+  }
+
+  sycl::free(A, q);
+
+  return 0;
+  // CHECK: addrspacecast ptr addrspace(4) %A to ptr addrspace(1)
```
Comment (sarnex) [inline sycl/test/invoke_simd/invoke_simd_address_space_inferral.cpp:75]: Maybe we could add an E2E test using Jason's perf test infrastructure to lock down the lower instruction count for a case that generates the runtime checks in IGC today

**Review by sarnex — COMMENTED** [review]
Nice job optimizing this case! First pass comments below

File: sycl/test/invoke_simd/invoke_simd_address_space_inferral.cpp [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+      cgh.parallel_for(
+          Range, [=](nd_item<1> item) [[intel::reqd_sub_group_size(VL)]] {
+            sycl::group<1> g = item.get_group();
+            sycl::sub_group sg = item.get_sub_group();
+
+            unsigned int i = g.get_group_id() * g.get_local_range() +
+                             sg.get_group_id() * sg.get_max_local_range();
+
+            invoke_simd(sg, SIMD_CALLEE1, uniform{A},
+                        uniform{LocalAcc.get_pointer().get()}, uniform{i});
```
Comment (v-klochkov) [inline sycl/test/invoke_simd/invoke_simd_address_space_inferral.cpp:62]: get_pointer() is deprecated API

File: llvm/lib/SYCLLowerIR/LowerInvokeSimd.cpp [hunk 4]
```diff
@@ -436,6 +495,7 @@ PreservedAnalyses SYCLLowerInvokeSimdPass::run(Module &M,
       ISCalls.insert(CI);
     }
   }
+
```
Comment (sarnex) [inline llvm/lib/SYCLLowerIR/LowerInvokeSimd.cpp:498]: nit: remove whitespace

File: sycl/test-e2e/ESIMD/PerformanceTests/invoke_simd_smoke.cpp [hunk 5]
```diff
@@ -0,0 +1,16 @@
+//==------------- invoke_simd_smoke.cpp  - DPC++ ESIMD on-device test----==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// REQUIRES: gpu
```
Comment (sarnex) [inline sycl/test-e2e/ESIMD/PerformanceTests/invoke_simd_smoke.cpp:8]: we probably want to restrict to a single GPU like we did for the rest of the perf tests

```
// REQUIRES: gpu-intel-dg2 && level_zero
```

File: sycl/test/invoke_simd/invoke_simd_address_space_inferral.cpp (hunk 2 above)
Comment (sarnex) [reply sycl/test/invoke_simd/invoke_simd_address_space_inferral.cpp:78]: yeah lets try, if it fails all the time we can remove it. thanks

**Review by sarnex — COMMENTED** [review]
lgtm, only minor comments, nice refactor!

File: sycl/test-e2e/ESIMD/PerformanceTests/invoke_simd_smoke.cpp (hunk 5 above)
Comment (sarnex) [reply sycl/test-e2e/ESIMD/PerformanceTests/invoke_simd_smoke.cpp:8]: yep, exactly

**Review by sarnex — APPROVED** [review]

**Comment by sarnex** [comment]
Fail not related

**Comment by sarnex** [comment]
@fineg74 It looks like the test is failing in poscommit, can you please take a look?

https://github.com/intel/llvm/actions/runs/9978520401/job/27584493411

**Comment by sarnex** [comment]
@fineg74 It builds the compiler using the same compiler version. Maybe build the compiler and then clone the same revision again and use that compiler to build it?

Also there is another postcommit fail [here](https://github.com/intel/llvm/actions/runs/9980297153/job/27586022526) so I am going to revert this PR for now.

**Comment by sarnex** [comment]
@fineg74 We can run postcommit in precommit by making a pull request to a new branch in intel/llvm you create that includes the `sycl-devops-pr` string

**Pull request merged**